rust/relay/ebpf-turn-router/src/main.rs

```diff
@@ -164,135 +164,661 @@ fn try_handle_turn_ipv4(ctx: &XdpContext, eth: Eth) -> Result<(), Error> {
+        number: channel_number.to_be_bytes(),
+        length: channel_data_length.to_be_bytes(),
+    };
 
-    let new_src = ipv4.dst(); // The IP we received the packet on will be the new source IP.
-    let new_dst = client_and_channel.client_ip();
-    let new_ipv4_total_len = ipv4.total_len() + CdHdr::LEN as u16;
+    Ok(())
+}
 
-    eth.swap_macs()?;
+#[inline(always)]
+fn try_handle_ipv4_channel_data_to_udp(ctx: &XdpContext) -> Result<(), Error> {
+    // SAFETY: XDP context provides valid packet memory. Bounds checked by ref_mut_at.
+    let ipv4 = unsafe { ref_mut_at::<Ipv4Hdr>(ctx, EthHdr::LEN)? };
+    // SAFETY: Writing to the new packet location after expansion.
+    // The ref_mut_at helper performs bounds checks before creating the reference.
+    // SAFETY: XDP context provides valid packet memory. Bounds checked by ref_mut_at.
+    let udp = unsafe { ref_mut_at::<UdpHdr>(ctx, EthHdr::LEN + Ipv4Hdr::LEN)? };
+    // SAFETY: XDP context provides valid packet memory. Bounds checked by ref_mut_at.
+    let cd = unsafe { ref_mut_at::<CdHdr>(ctx, EthHdr::LEN + Ipv4Hdr::LEN + UdpHdr::LEN)? };
+
+    let channel_number = u16::from_be_bytes(cd.number);
+
+    if !(CHAN_START..=CHAN_END).contains(&channel_number) {
+        return Err(Error::NotAChannelDataMessage);
+    }
 
-    let pseudo_header = ipv4.update(new_src, new_dst, new_ipv4_total_len);
+    let channel_data_len = u16::from_be_bytes(cd.length);
+    let expected_channel_data_len = udp.len() - UdpHdr::LEN as u16 - CdHdr::LEN as u16;
 
-    let udp_len = udp.len();
-    let new_udp_len = udp_len + CdHdr::LEN as u16;
+    // This can happen if we receive packets formed via GSO, like on a loopback interface.
+    if channel_data_len != expected_channel_data_len {
+        return Err(Error::BadChannelDataLength);
+    }
 
-    let channel_number = client_and_channel.channel();
-    let channel_data_length = udp_len - UdpHdr::LEN as u16; // The `length` field in the UDP header includes the header itself. For the channel-data field, we only want the length of the payload.
-
-    udp.update(
-        pseudo_header,
-        3478,
-        client_and_channel.client_port(),
-        new_udp_len,
-        channel_number,
-        channel_data_length,
+    let key = ClientAndChannelV4::new(ipv4.src_addr(), udp.source(), channel_number);
+
+    // SAFETY: We only write to these using a single thread in userspace.
+    if let Some(port_and_peer) = unsafe { CHAN_TO_UDP_44.get(&key) } {
+        // IPv4 to IPv4 - existing logic
+        handle_ipv4_to_ipv4_channel_to_udp(ctx, port_and_peer)?;
+        return Ok(());
+    }
+
+    // SAFETY: We only write to these using a single thread in userspace.
+    if let Some(port_and_peer) = unsafe { CHAN_TO_UDP_46.get(&key) } {
+        handle_ipv4_to_ipv6_channel_to_udp(ctx, port_and_peer)?;
+        return Ok(());
+    }
+
+    Err(Error::NoEntry(SupportedChannel::Chan4ToUdp))
+}
+
+#[inline(always)]
+fn handle_ipv4_to_ipv4_channel_to_udp(
```

Comment:
```suggestion
fn handle_ipv4_channel_to_ipv4_udp(
```

I think this naming would be a bit easier to understand?

Reply:
Yeah that could work. Generally when thinking through these you tend to think first in groups of whether the functions are cross-stack first, and then the lower layers. But I could see the case for this too.

I also want to add ASCII documentation to each of the eight functions showing the before and after packet header structure. It would have been super helpful to have this somewhere to reference as it applies to our specific situation.